README
olson-tools:  A slew of useful code created and gathered over the years.
    - Copyright 1998-2008 Spencer E. Olson (except as noted by individual files).


This software is released under the LGPL license except otherwise explicitly
stated in individual files included in this package.


components:
Jamroot         Overall Boost.Jam build file.
AUTHORS         A list of those who have contributed with contact info
ChangeLog       Brief descriptions of changes made between major revisions
CMakeLists.txt  The main (unfinished) CMake build script.
licenses        Copies of the licenses used for the pieces of this package.
docs            documentation, including doxygen generated api documentation
examples        Examples to demonstrate the use and capability of the package
INSTALL         Description of how one can install this package
README          Brief description of this package
src             source code
TODO            Things that I hope to implement someday


ChangeLog
commit 6fcf3b4c22de09f86fd0a5168ab780770ac87c33
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Wed Nov 4 14:36:26 2009 -0700

    added installation options

commit 835c722d0b6840cc6d7833dd3865b876a8225ef9
Author: Spencer Olson <olsonse@asok.hpmi.afrl.af.mil>
Date:   Wed Nov 4 10:02:27 2009 -0700

    fixed usage of va_start so that it is in line with the standard:  can't use reference variables

commit 1dd5df8a1a9260ac4167ad0dc1015b0eb9923215
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Thu Oct 29 12:10:27 2009 -0600

    added first cut of installation of libraries to a local stage directory

commit 65a3f215aaabf7c50dfbfa7bdeb66e9990e679e4
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Wed Oct 28 14:52:50 2009 -0600

    added some warning documentation to upper_triangular and a function to access edge length of nxn matrix

commit 927a1dced58d441668d01b4526e3ab9021e06305
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Tue Oct 27 13:58:39 2009 -0600

    small tweaks to remove compiler warnings....

commit bff625208e37a2f5228d8c474a41ff5de34d61d9
Author: U-OW\olsonse <olsonse@ow.(none)>
Date:   Tue Oct 27 09:26:25 2009 -0800

    Moved random stuff into olson_tools::random namespace when in c++.
    
    This fixed some build problems on Cygwin.  Changed other things to fix Cygwin
    build.
    
    Moved Timer.cpp and put its contents in a templated initialization variable in
    Timer.h.  Removed /olson-tools//timer library as Timer.h is now header only.

commit e5002d8662b620b529b60e087a7244de770d5e19
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Sat Oct 24 22:21:45 2009 -0600

    added /olson-tools//xml lib-alias to make /physical//runtime pulled in.

commit 718562716a67f30cce176c10c34edcf98c8e2ba9
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Tue Oct 20 22:15:16 2009 -0600

    added missing public declaration in upper_triangle

commit 112a8306d906e52522f5cce9d28f0b0020bb81f4
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Tue Oct 20 21:17:03 2009 -0600

    added docs for the upper_triangle class

commit c0800f0bd86e87a2ad085748406629d39d14f77d
Author: U-OW\olsonse <olsonse@ow.(none)>
Date:   Mon Oct 19 16:13:43 2009 -0700

    Fixed fast_log2 unit test for cygwin and updated for boost>1.34.
    
    /boost//unit_test_framework is also now linked statically

commit baa063e27fe004a35f61f7274df780eb688fa1f7
Author: Spencer Olson <olsonse@asok.hpmi.afrl.af.mil>
Date:   Fri Oct 16 13:41:28 2009 -0600

    Updated jam files in unit tests.  Fixed use of new olson_tools::V3 in unit tests.

commit 5c6385e8521f346d689fe980d318b1cd402fb692
Author: Spencer Olson <olsonse@asok.hpmi.afrl.af.mil>
Date:   Fri Oct 16 13:27:18 2009 -0600

    updated jam files to split up olson-tools into sub-libs and use boost correctly

commit f0dab4b2ff4083402ba953eecee409b4142a2e98
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Thu Oct 15 13:09:34 2009 -0600

    particledb->chimplib

commit a3db48ea00148d39cf8712cb5f213658051c8b30
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Thu Oct 15 12:43:02 2009 -0600

    Fixed jamfiles for testTrapfe and testXML

commit d8ac568589c19811ad5f95838b65dca172408d0c
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Sat Oct 3 14:24:55 2009 -0600

    removed old files from cvs that didn't get updated correctly

commit 64af3a9260a54bb232fe1415efa9a3946c05f939
Merge: 03a5c9d 6fb219b
Author: Spencer E. Olson <olsonse@olsonse-laptop.(none)>
Date:   Sat Oct 3 14:11:30 2009 -0600

    Merge branch 'master' of ../../git-svn/olson-tools into cvs-svn
    
    Conflicts:
    	examples/fast_pow/c/testfast_pow.c
    	examples/fast_pow/fortran/testfast_pow.F
    	examples/options/main.h
    	examples/trapfe/README

commit 6fb219bd0580e8544ec04a3300bbd999224a1e11
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>
Date:   Fri Oct 2 13:34:11 2009 +0000

    small fixes for ExecMerit and ExecFunc
    
    git-svn-id: svn+ssh://sftp.itd.umich.edu/afs/umich.edu/user/o/l/olsonse/src/svn-repository/trunk/olson-tools@843 e877f351-6dfb-0310-a246-97bd42f90100

commit bba3b00749e36b7026dd796b96962fc1527058fe
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>
Date:   Fri Oct 2 04:11:54 2009 +0000

    still modernizing ga internals
    
    git-svn-id: svn+ssh://sftp.itd.umich.edu/afs/umich.edu/user/o/l/olsonse/src/svn-repository/trunk/olson-tools@842 e877f351-6dfb-0310-a246-97bd42f90100

commit 9370d9c335c5e9e740576def725fa32eea85608c
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>
Date:   Thu Oct 1 21:51:02 2009 +0000

    tweaked log2 finding stuff when fast_log2 is not provided
    
    git-svn-id: svn+ssh://sftp.itd.umich.edu/afs/umich.edu/user/o/l/olsonse/src/svn-repository/trunk/olson-tools@841 e877f351-6dfb-0310-a246-97bd42f90100

commit 33c2e960fc960b95e25e15b3bb5fcf24994cb709
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>
Date:   Wed Sep 30 07:54:31 2009 +0000

    ExecMerit changed to work in new ga configuration
    
    git-svn-id: svn+ssh://sftp.itd.umich.edu/afs/umich.edu/user/o/l/olsonse/src/svn-repository/trunk/olson-tools@840 e877f351-6dfb-0310-a246-97bd42f90100

commit bd0c180d91fa6935ed6f1df441761edb761a60a3
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>
Date:   Wed Sep 30 07:34:33 2009 +0000

    Two types of changes:
    1.  PThreads stuff was improved in features.  Still need to show efficiency.
    
    2.  Genetic algorithm was revamped some more.  The Merit function is now a
        Functor class.  Updated examples to new usage.
    
    
    git-svn-id: svn+ssh://sftp.itd.umich.edu/afs/umich.edu/user/o/l/olsonse/src/svn-repository/trunk/olson-tools@839 e877f351-6dfb-0310-a246-97bd42f90100

commit 6c4d2f9b55f5b9acb7bf9656cafd7b0a6c7b8d1c
Author: olsonse <olsonse@e877f351-6dfb-0310-a246-97bd42f90100>

src/util.C
// -*- c++ -*-
// $Id: util.C,v 1.1.1.1 2005/01/08 04:27:25 olsonse Exp $
/*
 * Copyright 1998-2004 Spencer Eugene Olson --- All Rights Reserved
 *
 * $Log: util.C,v $
 * Revision 1.1.1.1  2005/01/08 04:27:25  olsonse
 * Initial import
 *
 * Revision 1.1.1.1  2003/05/18 02:05:10  olsonse
 * Initial import of atomguide into cvs.
 * This marks the change from a running version that did not really do cells
 * correctly, to one where cells will correctly be done (hopefully).
 *
 * Revision 1.7  2001/01/11 16:16:34  olsons
 * Improved plotting:  finished 3d plotting, added ability to change x-display.
 *
 * Revision 1.6  2000/10/25 18:19:46  olsons
 * Finished (almost) all of the remaining functions and menus that were missing.
 *
 * Revision 1.5  2000/06/15 19:32:56  olsons
 * Added get_Long_Comment function.
 *
 * Revision 1.4  2000/05/06 20:39:26  olsons
 * New interface design for next version of xuvmulti.
 *
 * Revision 1.3  1999/08/18 19:35:39  olsons
 * Changed util.C: getMat so that you can enter either material name or number.
 *
 * Revision 1.2  1999/08/18 13:50:41  olsons
 * *** empty log message ***
 *
 * Revision 1.1  1999/07/03 05:29:37  olsons
 * Added character-based user interface to XUVMulti
 *
 *
 */


/* TextUI Library
 * by Spencer Olson
 * copyright 1999
 *
 * Free use granted as long as this notice is included.
*/


#include "util.h"
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <string>
#include <iostream>
#include <cfloat>
#include <stdlib.h>

using namespace std;

#define GNU_READLINE
#ifdef GNU_READLINE

extern "C" {
	char * readline( const char * );
        int add_history( const char * );
        char * rl_readline_name;
	int rl_stuff_char (int c);
}

void rl_stuff( const char * str ) {
  for( int i = 0, len = strlen(str); i < len; rl_stuff_char( str[i++] ) );
}//rl_stuff

#endif

void text_getMinDoubleMax(double &min,double &val,double &max,
                  bool &stat, const string & qstring ) {
  // reads from stdin to find the max & min and value of a
  // certain parameter named parname. 

  string tmpqstr = string( "Do you want the parameter \'" )
                   + qstring + string( "\' to be static? ");

  double tmpmin(-1e10),tmpmax(1e10);
  if( ! ( stat = text_getBool( stat, tmpqstr ) ) ) {
    while(1) {
      tmpmin=text_getDouble(-1e10,min,1e10,"Minimum Value: ");
      tmpmax=text_getDouble(-1e10,max,1e10,"Maximum Value: ");
      if(tmpmin<=tmpmax) {min=tmpmin;max=tmpmax;break;}
      cout << "Error: minimum value greater than maximum value.\n";
    }//while
  }//if not static
  val = text_getDouble( tmpmin, val, tmpmax, qstring );
}//getMinDoubleMax

void text_getMinIntMax(int &min,int &val,int &max,
                  bool &stat, const string & qstring ) {
  // reads from stdin to find the max & min and value of a
  // certain parameter named parname. 

  string tmpqstr = string( "Do you want the parameter \'" )
                   + qstring + string( "\' to be static? ");

  int tmpmin(-60000),tmpmax(60000);
  if( !( stat = text_getBool( stat, tmpqstr ) ) ) {
    while(1) {
      tmpmin = text_getInt(-60000,min,60000,"Minimum Value: ");
      tmpmax = text_getInt(-60000,max,60000,"Maximum Value: ");
      if(tmpmin<=tmpmax) {min=tmpmin;max=tmpmax;break;}
      cout<<"Error: minimum value greater than maximum value.\n";
    }//while
  }//if not static
  val = text_getInt( tmpmin, val, tmpmax, qstring );
}//getMinIntMax

void text_getMinFloatMax( float &min, float &val, float &max,
                   bool &stat, const string & qstring ) {
  // reads from stdin to find the max & min and value of a
  // certain parameter named parname. 

  string tmpqstr = string( "Do you want the parameter \'" )
                   + qstring + string( "\' to be static? ");

  float tmpmin(-1e10),tmpmax(1e10);
  if( !( stat = text_getBool( stat, tmpqstr ) ) ) {
    while(1) {
      tmpmin=text_getFloat(float(-1e10),min,float(1e10),"Minimum Value: ");
      tmpmax=text_getFloat(float(-1e10),max,float(1e10),"Maximum Value: ");
      if(tmpmin<=tmpmax) {min=tmpmin;max=tmpmax;break;}
      cout<<"Error: minimum value greater than maximum value.\n";
    }//while
  }//if not static
  val=text_getFloat(tmpmin,val,tmpmax,qstring);
}//getMinFloatMax

double text_getDouble( double min, double defaultval, double max,
               const string & qstring) {
  //query the user for some specific value
  char buf[100];
  double selection=-FLT_MAX;
  while((selection<min)||(selection>max)) {
    cout<<qstring<<"["<<defaultval<<"] ";
    cin.getline(buf,100);
    if(strlen(buf)==0 &&(defaultval>=min)&&(defaultval<=max)) return defaultval;
    else if(sscanf(buf,"%lf",&selection)
       &&(selection>=min)&&(selection<=max)) {
       return selection;
    }//if
    selection=-FLT_MAX;
    cout<<"Value must be between "<<min<<" and "<<max<<".\n";
  }//while
  return 0;//this line of code should never be reached
}//getDouble

int text_getInt( int min, int defaultval, int max, const string & qstring ) {
  //query the user for some specific value
  char buf[100];
  int selection=min-1;
  while((selection<min)||(selection>max)) {
    cout<<qstring<<"["<<defaultval<<"] ";
    cin.getline(buf,100);
    if(strlen(buf)==0) sprintf(buf,"%d",defaultval);
    if(sscanf(buf,"%d",&selection)&&(selection>=min)&&(selection<=max))
       return selection;
    selection=min-1;
    cout<<"Value must be between "<<min<<" and "<<max<<".\n";
  }//while
  return 0;//this line of code should never be reached
}//getInt

float text_getFloat(float min, float defaultval, float max, const string & qstring) {
  //query the user for some specific value
  char buf[100];
  float selection=-FLT_MAX;
  while((selection<min)||(selection>max)) {
    cout<<qstring<<"["<<defaultval<<"] ";
    cin.getline(buf,100);
    if(strlen(buf)==0) sprintf(buf,"%f",defaultval);
    if(sscanf(buf,"%f",&selection)
       &&(selection>=min)&&(selection<=max)) {
       return selection;
    }//if
    selection=-FLT_MAX;
    cout<<"Value must be between "<<min<<" and "<<max<<".\n";
  }//while
  return 0;//this line of code should never be reached
}//getFloat

bool text_getBool( bool defaultval, const string & qstring ) {
  //ask a y/n question
  char selection[100];selection[0]='\0';
  char defchar; if(defaultval)defchar='y'; else defchar='n';
  while(selection[0]=='\0') {
    cout<<qstring<<"["<<defchar<<"] ";
    cin.getline(selection,100);
    if(strlen(selection)==0) return defaultval;
    if((selection[0]=='y')||(selection[0]=='Y'))return true;
    else if((selection[0]=='n')||(selection[0]=='N'))return false;
    selection[0]='\0';
    cout<<"You must enter y or n.\n";
  }//while
  return false;//this line of code should never be reached
}//getBool

int text_menu(int nitems, const char * items[], const string & title ) {
  //returns the selection
  int i,selection=0;
  static char buf[100];
  while(selection==0) {
    cout<<title<<endl;
    for(i=1;i<=nitems;i++) cout<<"\t"<<i<<".\t"<<items[i-1]<<endl;
    cout<<"\t\tYour Choice: ";
    cin.getline(buf,100);
    sscanf(buf,"%d",&selection);
    if((selection>0)&&(selection<=nitems)) 
      return selection-1;//-1 because we will return index of option
    selection=0;
    cout<<"Please select one of the following items"
        <<" by entering the adjacent index.\n";
  }//while
  return selection;//this line of code never should be reached
}//text_menu

int text_menu( int nitems, const char * items[], bool enabled[], const string & title ) {
  //returns the selection
  int j, msize, selection;
  char ** enableditems = new char*[nitems];

  for( j = 0, msize = 0; j < nitems; j++ ) {
    if( enabled[ j ] ) {
      enableditems[ msize ] = new char[ strlen( items[ j ] ) + 1 ];
      strcpy( enableditems[ msize++ ], items[ j ] );
    }//
  }//for

  selection = text_menu( msize, (const char **)enableditems, title );

  for(j=0;selection>=0;j++) {
    if(enabled[j]) selection--;
  }//for
  selection = j-1;

  for( j = 0; j < msize; delete[] enableditems[j++] );
  delete[] enableditems;
  return selection;
}//text_menu

string & text_getStr(string & str, const string & qstring ) {
  // a pretty trivial function which just gets a string from the user 
#ifndef GNU_READLINE
  string buf;
  cout << qstring << "[" << str << "] ";
  getline( cin, buf );
  if( buf.length () != 0 ) str = buf;
#else
  char * buf;
  rl_stuff ( str.c_str() );
  buf = readline( qstring.c_str() );
  if( buf ) {
    add_history( buf ), str = buf;
    free(buf);
  }//if
  else str = "";
#endif
  return str;
}//getStr

void text_alert( const string & message ) {
  cout << message << endl;
  pausetext();
}//text_alert

void text_alert( const char * message ) {
  cout << message << endl;
  pausetext();
}//text_alert

int pausetext(){
  char c;
  cout<<"Press <Enter> to continue.";
  cout.flush();
  cin.get( c );
  return int( c );
} // pausetext

string & text_get_Filename( string & defaultval, const string & prompt ){
  return text_getStr( defaultval, prompt );
} // text_get_Filename

string & text_get_Long_Comment( string & str, const string & prompt ){
  // a pretty trivial function which just gets a string from the user 
  string buf;
  cout << prompt << endl << "[" << str << "]" << endl
       << "Please leave a blank line to end (or press return now to keep previous value)"
       << endl;
  getline( cin, buf );
  if( buf.length () != 0 ) {
    str = buf;
    getline( cin, buf );
    while( buf.length () != 0 ) {
      str += '\n';
      str += buf;
      getline( cin, buf );
    }//while
  }//if
  return str;
} // text_get_Long_Comment

